aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -211,68 +213,56 @@ impl<'r, 'l> SessionExt<'r, 'l> {
+                    resources_filtered.insert(struct_tag, blob_op);
                 }
             }
 
-            for (name, blob_op) in modules {
-                change_set_filtered
-                    .add_module_op(ModuleId::new(addr, name), blob_op)
-                    .map_err(|_| common_error.clone())?;
-            }
+            change_set_filtered.add_account_changeset(addr, AccountChangeSet::from_modules_resources(modules, resources_filtered)).map_err(|_| common_error())?;
 
             for (resource_tag, resources) in resource_groups {
-                let source_data = remote
-                    .get_resource_group_data(&addr, &resource_tag)
-                    .map_err(|_| common_error.clone())?;
-                let (mut source_data, create) = if let Some(source_data) = source_data {
-                    let source_data =
-                        bcs::from_bytes(&source_data).map_err(|_| common_error.clone())?;
-                    (source_data, false)
-                } else {
-                    (BTreeMap::new(), true)
-                };
+                let mut create = false;
+                let mut source_data = remote
+                    .get_remaining_resource_group(&addr, &resource_tag)
+                    .unwrap_or_else(|| {
+                        create = true;
+                        BTreeMap::new()
+                    });
 
                 for (struct_tag, current_op) in resources.into_resources() {
+                    if source_data.contains_key(&struct_tag) {
+                        return Err(common_error());
+                    }
                     match current_op {
                         MoveStorageOp::Delete => {
-                            source_data
-                                .remove(&struct_tag)
-                                .ok_or_else(|| common_error.clone())?;
+                            let data = source_data.remove(&struct_tag);
+                            debug_assert!(data.is_some());
                         },
-                        MoveStorageOp::Modify(new_data) => {
-                            let data = source_data
-                                .get_mut(&struct_tag)
-                                .ok_or_else(|| common_error.clone())?;
-                            *data = new_data;
+                        MoveStorageOp::Modify(data) => {
+                            let data = source_data.insert(struct_tag, data);
+                            debug_assert!(data.is_some());
                         },
                         MoveStorageOp::New(data) => {
                             let data = source_data.insert(struct_tag, data);
-                            if data.is_some() {
-                                return Err(common_error);
-                            }
-                        },
-                    }
+                            debug_assert!(data.is_none());
+                        },                    }
```

Comment:
The comment dropped, because the code changed, but this does cause behavioral changes. Hiding or delaying errors doesn't seem right here.